syntax = "proto3";
option java_package = "io.agenc.externalframework";

message DataRow {
    repeated double data = 1;
}

message ColumnMetaData {
    string name = 1;
    double lowerBound = 2;
    double upperBound = 3;
    ColumnType type = 4;
}

enum ColumnType {
    COLUMNTYPE_UNDEFINED = 0;
    COLUMNTYPE_INPUT = 1;
    COLUMNTYPE_TARGET = 2;
}

message DataPackage {
    string name = 1;
    
    repeated ColumnMetaData metaData = 2;
    repeated DataRow data = 3;
}

message LearnerData {
    int32 returnCode = 1;
    optional string resultMessage = 2;

    int32 identifier = 3;
}

service ExternalLearner {
    rpc TrainLearner(DataPackage) returns (LearnerData) {}

    rpc RunLearner(DataRow) returns (DataRow) {}
}

